### PR TITLE
test(spanner): make use of the IsOkAndHolds() StatusOr<T> matcher

### PR DESCRIPTION
**Pull request by devbww**
This started out as an exercise to use IsOkAndHolds(m) within the Spanner tests, just to make sure it was working as intended, but soon turned into a cleanup for other assert/expect mismatches.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11262)

### CODE REVIEW

File: google/cloud/spanner/client_test.cc [hunk 1]
```diff
@@ -1056,8 +1057,8 @@ TEST(ClientTest, ProfileQueryWithOptionsSuccess) {
   };
   int row_number = 0;
   for (auto& row : StreamOf<RowType>(rows)) {
-    EXPECT_STATUS_OK(row);
-    EXPECT_EQ(*row, expected[row_number]);
+    ASSERT_LT(row_number, expected.size());
```
Comment (coryan) [inline google/cloud/spanner/client_test.cc:1060]: Another opportunity for `ElementsAre(IsOkAndHolds(...`

**Review by coryan — APPROVED** [review]

File: google/cloud/spanner/client_test.cc [hunk 2]
```diff
@@ -204,8 +205,8 @@ TEST(ClientTest, ExecuteQuerySuccess) {
   };
   int row_number = 0;
```
[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/spanner/client_test.cc]: Done.

File: google/cloud/spanner/client_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/spanner/client_test.cc]: Done.

File: google/cloud/spanner/integration_tests/data_types_integration_test.cc [hunk 3]
```diff
@@ -537,7 +519,7 @@ TEST_F(PgDataTypeIntegrationTest, InsertAndQueryWithJson) {
         if (!dml_result) return dml_result.status();
         return Mutations{};
       });
-  ASSERT_STATUS_OK(commit_result);
+  EXPECT_STATUS_OK(commit_result);
```
[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/spanner/integration_tests/data_types_integration_test.cc]: Done.

File: google/cloud/spanner/integration_tests/data_types_integration_test.cc [hunk 4]
```diff
@@ -778,7 +760,7 @@ TEST_F(DataTypeIntegrationTest, InsertAndQueryWithStruct) {
   ASSERT_STATUS_OK(row);
 
   auto const& v = std::get<0>(*row);
-  EXPECT_EQ(1, v.size());
+  ASSERT_EQ(1, v.size());
```
[the remark this answers is not in the document]
Comment (devbww) [reply google/cloud/spanner/integration_tests/data_types_integration_test.cc]: Done.

File: google/cloud/spanner/client_test.cc [hunk 5]
```diff
@@ -985,25 +973,18 @@ TEST(ClientTest, ProfileQuerySuccess) {
...
   ASSERT_TRUE(actual_plan);
   EXPECT_THAT(*actual_plan, IsProtoEqual(stats.query_plan()));
 
   auto actual_stats = rows.ExecutionStats();
   ASSERT_TRUE(actual_stats);
-  std::unordered_map<std::string, std::string> expected_stats{
-      {"elapsed_time", "42 secs"}};
-  EXPECT_EQ(expected_stats, *actual_stats);
+  EXPECT_THAT(*actual_stats,
+              UnorderedElementsAre(Pair("elapsed_time", "42 secs")));
```
Comment (coryan) [inline google/cloud/spanner/client_test.cc:987]: I assume you have a reason to prefer `UnorderedElementsAre()` over `ElementsAre()` even though there is a single element?  If you have a minute can you share why?

**Review by coryan — APPROVED** [review]
Thanks, very cool cleanups.

File: google/cloud/spanner/client_test.cc (hunk 5 above)
Comment (devbww) [reply google/cloud/spanner/client_test.cc:987]: Yes ... what Darren said.

**Pull request merged**